io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -359,9 +357,22 @@ liftST :: StrictST.ST s a -> SimM s a
 liftST action = SimM $ \k -> LiftST action k
 
 instance MonadTime (SimM s) where
-  type Time (SimM s) = VTime
 
-  getMonotonicTime = SimM $ \k -> GetTime k
+  getMonotonicTime = SimM $ \k -> GetMonoTime k
+  getCurrentTime   = SimM $ \k -> GetWallTime k
+
+-- | Set the current wall clock time for the thread's clock domain.
+--
+setCurrentTime :: UTCTime -> SimM s ()
+setCurrentTime t = SimM $ \k -> SetWallTime t (k ())
+
+-- | Put the thread into a new wall clock domain, not shared with the parent
+-- thread. Changing the wall clock time in the new clock domain will not affect
+-- the other clock of other threads. All threads forked by this thread from
+-- this point onwards will share the new clock domain.
+--
+unshareClock :: SimM s ()
+unshareClock = SimM $ \k -> UnshareClock (k ())
```

Comment:
@nfrisby I'm thinking on how to integrate it in the protocol tests.

First of all, I think some code will no longer be parametric in `IOLike m => m` but will need to be instantiated with `forall s. SimM s`. Alternatively, we introduce a type class for this that only has an instance for the simulator. The next step is to spawn each node in its own thread, because I don't think they currently are (each node itself spawns a bunch threads). Finally, there will be some plumbing of generated test input (per node wall clock domain) to `createNode`. What do you think?

Reply:
Yes, the top level test setup would need to spawn a thread per host, and then each host's main thread does the `unshareClock` and `setCurrentTime` to get an offset, followed by creating the node (so all child threads live in the same clock domain).

That top level test setup code will of course no longer be fully polymorphic as it's specific to the simulator.